src/fomo.jl

```diff
@@ -0,0 +1,399 @@
+Changing any of the input arguments will affect the subsequent iterations.
+In particular, setting `stats.status = :user` will stop the algorithm.
+All relevant information should be available in `nlp` and `solver`.
+Notably, you can access, and modify, the following:
+- `solver.x`: current iterate;
+- `solver.gx`: current gradient;
+- `stats`: structure holding the output of the algorithm (`GenericExecutionStats`), which contains, among other things:
+  - `stats.dual_feas`: norm of current gradient;
+  - `stats.iter`: current iteration counter;
+  - `stats.objective`: current objective function value;
+  - `stats.status`: current status of the algorithm. Should be `:unknown` unless the algorithm has attained a stopping criterion. Changing this to anything will stop the algorithm, but you should use `:user` to properly indicate the intention.
+  - `stats.elapsed_time`: elapsed time in seconds.
+
+# Examples
+## `fomo`
+```jldoctest
+using JSOSolvers, ADNLPModels
+nlp = ADNLPModel(x -> sum(x.^2), ones(3))
+stats = fomo(nlp)
+
+# output
+
+"Execution stats: first-order stationary"
+```
+
+```jldoctest
+using JSOSolvers, ADNLPModels
+nlp = ADNLPModel(x -> sum(x.^2), ones(3))
+solver = FomoSolver(nlp);
+stats = solve!(solver, nlp)
+
+# output
+
+"Execution stats: first-order stationary"
+```
+## `R2`
+```jldoctest
+using JSOSolvers, ADNLPModels
+nlp = ADNLPModel(x -> sum(x.^2), ones(3))
+stats = R2(nlp)
+
+# output
+
+"Execution stats: first-order stationary"
+```
+
+```jldoctest
+using JSOSolvers, ADNLPModels
+nlp = ADNLPModel(x -> sum(x.^2), ones(3))
+solver = FoSolver(nlp);
+stats = solve!(solver, nlp)
+
+# output
+
+"Execution stats: first-order stationary"
+```
+"""
+mutable struct FomoSolver{T, V} <: AbstractFirstOrderSolver
+  x::V
+  g::V
+  c::V
+  m::V
+  d::V
+  p::V
+  α::T
+end
+
+function FomoSolver(nlp::AbstractNLPModel{T, V}) where {T, V}
+  x = similar(nlp.meta.x0)
+  g = similar(nlp.meta.x0)
+  c = similar(nlp.meta.x0)
+  m = fill!(similar(nlp.meta.x0), 0)
+  d = fill!(similar(nlp.meta.x0), 0)
+  p = similar(nlp.meta.x0)
+  return FomoSolver{T, V}(x, g, c, m, d, p, T(0))
+end
+
+@doc (@doc FomoSolver) function fomo(nlp::AbstractNLPModel{T, V}; kwargs...) where {T, V}
+  solver = FomoSolver(nlp)
+  solver_specific = Dict(:avgβmax => T(0.))
+  stats = GenericExecutionStats(nlp;solver_specific=solver_specific)
+  return solve!(solver, nlp, stats; kwargs...)
+end
+
+function SolverCore.reset!(solver::FomoSolver{T}) where {T}
+  fill!(solver.m,0)
+  solver
+end
+
+SolverCore.reset!(solver::FomoSolver, ::AbstractNLPModel) = reset!(solver)
+
+mutable struct FoSolver{T, V} <: AbstractFirstOrderSolver
+  x::V
+  g::V
+  c::V
+  α::T
+end
+
+function FoSolver(nlp::AbstractNLPModel{T, V}) where {T, V}
+  x = similar(nlp.meta.x0)
+  g = similar(nlp.meta.x0)
+  c = similar(nlp.meta.x0)
+  return FoSolver{T, V}(x, g, c, T(0))
+end
+
+@doc (@doc FomoSolver) function R2(nlp::AbstractNLPModel{T, V}; kwargs...) where {T, V}
```

Comment:
It’s odd that the solver object is called `FoSolver` but the solver is called `R2`. Call the solver `fo` and define `R2 = fo` with `r2_step()` and `TR = fo` with `tr_step()`.